local.go

```diff
@@ -54,59 +54,100 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
 	}
 
 	startTime := time.Now()
-
 	funcs, funcInfo := getFunctionInfo(f)
+	status := NotTerminated
+
+	if funcInfo.IsStatuser {
+		_, err := funcs.status.Status()
+		if err != nil {
+			// Function status is bad, panic.
+			// TODO: Document this panic in the function comment.
+			panic(fmt.Sprintf("bad initial function status: %v", err))
+		}
+	}
 
-	if method == nil {
-		method = getDefaultMethod(funcInfo)
+	if settings.Recorder != nil {
+		// Initialize Recorder first. If it fails, we avoid the (possibly
+		// time-consuming) evaluation of F and DF at the starting location.
+		err := settings.Recorder.Init(funcInfo)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	if settings == nil {
 		settings = DefaultSettings()
 	}
-
 	stats := &Stats{}
-	location, err := getStartingLocation(funcs, funcInfo, initX, stats, settings)
+
+	optLoc, evalType, err := getStartingLocation(funcs, funcInfo, initX, stats, settings)
 	if err != nil {
 		return nil, err
 	}
 
-	optLoc := &Location{}
-	// update stats (grad norm, function value, etc.) so that things are
-	// initialized for the first convergence check
-	update(location, optLoc, stats, NoIteration, startTime)
-
+	// Runtime is the only Stats field that needs to be updated here.
+	stats.Runtime = time.Since(startTime)
+	// Send optLoc to Recorder before checking it for convergence.
 	if settings.Recorder != nil {
-		err = settings.Recorder.Init(funcInfo)
+		// TODO(vladimir-ch): Replace NoIteration with InitialIteration when it
+		// is added.
+		err = settings.Recorder.Record(optLoc, evalType, NoIteration, stats)
 		if err != nil {
-			return &Result{Status: Failure}, err
+			goto Cleanup
 		}
 	}
 
-	// actually perform the minimization
-	status, err := minimize(settings, location, method, funcInfo, stats, funcs, optLoc, startTime)
+	// Check if the starting location satisfies the convergence criteria.
+	// TODO(vladimir-ch): Replace MajorIteration with InitialIteration when it
+	// is added and checkConvergence() updated to handle it.
+	status = checkConvergence(optLoc, MajorIteration, stats, settings)
+	if status == NotTerminated {
+		// The starting location is not good enough, we need to start the
+		// minimization run.
+
+		// First, check the status of the function to see if it is able to
+		// evaluate.
+		// TODO(vladimir-ch): Is this the best place for checking the function
+		// status? Revise after minimize() is updated.
+		if funcInfo.IsStatuser {
```

Comment:
Is this check actually necessary? It already gets checked before the next function evaluation in minimize. Removing it would make this much simpler and would also allow the removal of the Cleanup label (sorry for not seeing this before).